src/node/txdownloadman_impl.cpp

```diff
@@ -286,4 +288,160 @@ void TxDownloadManagerImpl::MempoolAcceptedTx(const CTransactionRef& tx)
     // If it came from the orphanage, remove it. No-op if the tx is not in txorphanage.
     m_orphanage.EraseTx(tx->GetWitnessHash());
 }
+
+node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransactionRef& ptx, const TxValidationState& state, NodeId nodeid, bool first_time_failure)
+{
+    const CTransaction& tx{*ptx};
+    // Results returned to caller
+    // Whether we should call AddToCompactExtraTransactions at the end
+    bool add_extra_compact_tx{first_time_failure};
+    // Hashes to pass to AddKnownTx later
+    std::vector<uint256> unique_parents;
+    // Populated if failure is reconsiderable and eligible package is found.
+    std::optional<node::PackageToValidate> package_to_validate;
+
+    if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
+        // Only process a new orphan if this is a first time failure, as otherwise it must be either
+        // already in orphanage or from 1p1c processing.
+        if (first_time_failure) {
+            bool fRejectedParents = false; // It may be the case that the orphans parents have all been rejected
+
+            // Deduplicate parent txids, so that we don't have to loop over
+            // the same parent txid more than once down below.
+            unique_parents.reserve(tx.vin.size());
+            for (const CTxIn& txin : tx.vin) {
+                // We start with all parents, and then remove duplicates below.
+                unique_parents.push_back(txin.prevout.hash);
+            }
+            std::sort(unique_parents.begin(), unique_parents.end());
+            unique_parents.erase(std::unique(unique_parents.begin(), unique_parents.end()), unique_parents.end());
+
+            // Distinguish between parents in m_lazy_recent_rejects and m_lazy_recent_rejects_reconsiderable.
+            // We can tolerate having up to 1 parent in m_lazy_recent_rejects_reconsiderable since we
+            // submit 1p1c packages. However, fail immediately if any are in m_lazy_recent_rejects.
+            std::optional<uint256> rejected_parent_reconsiderable;
+            for (const uint256& parent_txid : unique_parents) {
+                if (RecentRejectsFilter().contains(parent_txid)) {
+                    fRejectedParents = true;
+                    break;
+                } else if (RecentRejectsReconsiderableFilter().contains(parent_txid) &&
+                           !m_opts.m_mempool.exists(GenTxid::Txid(parent_txid))) {
+                    // More than 1 parent in m_lazy_recent_rejects_reconsiderable: 1p1c will not be
+                    // sufficient to accept this package, so just give up here.
+                    if (rejected_parent_reconsiderable.has_value()) {
+                        fRejectedParents = true;
+                        break;
+                    }
+                }
+                rejected_parent_reconsiderable = parent_txid;
```

Comment:
I think this line got pulled out of the `else if` conditional. Won't this set the value for all unique parents and reject any transaction with two unique prevout hashes?

Reply:
Ugh yes, a bad rebase sorry. I've been using a new setup and haven't perfected my diff-viewing yet.

I don't understand how 0 tests failed, though 😬 gotta look into that

Reply:
> I don't understand how 0 tests failed, though 😬 gotta look into that

yeah was a bit worrying to me too... glad we can add unit tests in this PR!